components/Plate/plugins/concept.jsx

```diff
@@ -0,0 +1,118 @@
+import { Transforms, Text, Node, Editor as SlateEditor } from 'slate';
+import { createPluginFactory } from "@udecode/plate";
+
+import { ELEMENT_CONCEPT } from "../../../utils/slate";
+
+const conceptRegex = /\[\[(.*)\]\](.*)/
+
+function hasConceptParent(editor, path) {
+  const parent = Node.get(editor, path.slice(0, -1))
+  if (parent.type === ELEMENT_CONCEPT) {
+    return true
+  } else {
+    return false
+  }
+}
+
+export const withConcepts = editor => {
+  const { normalizeNode } = editor
+
+  editor.normalizeNode = entry => {
+    const [node, path] = entry
+    if (Text.isText(node) && !hasConceptParent(editor, path)) {
+      const conceptMatch = node.text.match(conceptRegex)
+      if (conceptMatch) {
+        const { index, 0: match, 1: name } = conceptMatch
+        const at = { anchor: { path, offset: index }, focus: { path, offset: index + match.length } }
+        Transforms.wrapNodes(editor, { type: ELEMENT_CONCEPT, children: [] }, { at, split: true })
+        return
+      }
+    } else if (node.type === ELEMENT_CONCEPT) {
+
+      // Migrate from old to new concept format
+      if (node.value) {
+        Transforms.setNodes(editor, {
+          name: node.value
+        }, { at: path })
+        const childText = `[[${node.value}]]`
+        if (node.children[0].text != childText) {
+          const childTextStart = { path: [...path, 0], offset: 0 }
+          const lastChildIndex = (node.children.length - 1)
+          const lastTextPositionIndex = node.children[lastChildIndex].text.length
+          const childTextEnd = { path: [...path, lastChildIndex], offset: lastTextPositionIndex }
+          const childTextRange = { anchor: childTextStart, focus: childTextEnd }
+          Transforms.insertText(editor, childText, { at: childTextRange })
+        }
+        Transforms.unsetNodes(editor, 'value', { at: path })
+        return
+      }
+
+      const conceptMatch = Node.string(node).match(conceptRegex)
+      if (conceptMatch) {
+        const [_, name, extra] = conceptMatch
+
+        // make sure name always matches text
+        if (node.name !== name) {
+          Transforms.setNodes(editor, { name }, { at: path })
+          return
+        }
+
+        // make sure a concept doesn't eat the text after it
+        if (extra && extra !== '') {
+          const childText = `[[${node.name}]]`
+          Transforms.splitNodes(editor, {
+            at: { path: [...path, 0], offset: childText.length },
+            match: n => (n.type === ELEMENT_CONCEPT)
+          })
+          return
+        }
+      } else {
+        Transforms.unwrapNodes(editor, { match: n => n.type === ELEMENT_CONCEPT })
+        return
+      }
+    }
+
+    normalizeNode(entry)
+  }
+
+  return editor
+}
+
+export const LEAF_CONCEPT_START = 'conceptLeafStart'
+export const LEAF_CONCEPT_END = 'conceptLeafEnd'
```

Comment:
Why do we need a start node and and end node? Would you mind adding a comment explaining it? It makes sense that we would need to do something like this, it's just not clear how this is working to me at first glance. It looks like an ELEMENT_CONCEPT includes a leaf start, a leaf end, and then presumably a text node in the middle with the name?

Reply:
yea I'll add a comment here - basically we define "decorators" for the start and end of a concept and then provide a React component to render each of them

Reply:
done!